File: backend/utils/toon_parser.py
"""TOON (Token-Oriented Object Notation) parser utilities."""

from typing import Dict, Any


def parse_toon(toon_string: str) -> Dict[str, Any]:
    """
    Parse TOON format string to dictionary.
    
    TOON format uses pipe (|) for field separation and colon (:) for key-value pairs.
    Example: "vuln:SQL_INJ|sev:HIGH|file:auth.py|ln:45-47|type:unsanitized_input"
    
    Args:
        toon_string: String in TOON format
        
    Returns:
        Dictionary representation of the TOON data
        
    Example:
        >>> parse_toon("vuln:SQL_INJ|sev:HIGH|file:auth.py")
        {'vuln': 'SQL_INJ', 'sev': 'HIGH', 'file': 'auth.py'}
    """
    if not toon_string or not isinstance(toon_string, str):
        return {}
    
    result = {}
    
    # Split by pipe delimiter
    fields = toon_string.split("|")
    
    for field in fields:
        field = field.strip()
        if not field:
            continue
        
        # Split by colon to get key-value pairs
        if ":" in field:
            key, value = field.split(":", 1)
            result[key.strip()] = value.strip()
    
    return result


def to_toon(data: Dict[str, Any]) -> str:
    """
    Convert dictionary to TOON format string.
    
    Args:
        data: Dictionary to convert
        
    Returns:
        String in TOON format
        
    Example:
        >>> to_toon({'vuln': 'SQL_INJ', 'sev': 'HIGH', 'file': 'auth.py'})
        'vuln:SQL_INJ|sev:HIGH|file:auth.py'
    """
    if not data or not isinstance(data, dict):
        return ""
    
    # Convert dict to TOON format
    fields = []
    for key, value in data.items():
        # Convert value to string
        value_str = str(value)
        fields.append(f"{key}:{value_str}")
    
    return "|".join(fields)


def parse_toon_array(toon_array: str) -> list[Dict[str, Any]]:
    """
    Parse an array of TOON strings (comma-separated).
    
    Args:
        toon_array: Comma-separated TOON strings
        
    Returns:
        List of dictionaries
        
    Example:
        >>> parse_toon_array("vuln:XSS|sev:HIGH,vuln:SQL_INJ|sev:CRIT")
        [{'vuln': 'XSS', 'sev': 'HIGH'}, {'vuln': 'SQL_INJ', 'sev': 'CRIT'}]
    """
    if not toon_array:
        return []
    
    # Split by comma and parse each TOON string
    toon_strings = toon_array.split(",")
    return [parse_toon(toon_str.strip()) for toon_str in toon_strings if toon_str.strip()]


def to_toon_array(data_list: list[Dict[str, Any]]) -> str:
    """
    Convert list of dictionaries to comma-separated TOON strings.
    
    Args:
        data_list: List of dictionaries
        
    Returns:
        Comma-separated TOON strings
        
    Example:
        >>> to_toon_array([{'vuln': 'XSS', 'sev': 'HIGH'}, {'vuln': 'SQL_INJ', 'sev': 'CRIT'}])
        'vuln:XSS|sev:HIGH,vuln:SQL_INJ|sev:CRIT'
    """
    if not data_list:
        return ""
    
    return ",".join([to_toon(item) for item in data_list])

File: frontend/lib/toon-parser.ts
// TOON (Token-Oriented Object Notation) Parser
// Format: key1:value1|key2:value2|key3:value3

import { ParsedVulnerability } from '@/types/scan';

/**
 * Parse a TOON format string into a key-value object
 * @param toonString - String in format "key1:value1|key2:value2"
 * @returns Object with parsed key-value pairs
 */
export function parseTOON(toonString: string): Record<string, string> {
  if (!toonString || typeof toonString !== 'string') {
    return {};
  }

  const result: Record<string, string> = {};
  const fields = toonString.split('|');

  fields.forEach((field) => {
    const colonIndex = field.indexOf(':');
    if (colonIndex === -1) return;

    const key = field.substring(0, colonIndex).trim();
    const value = field.substring(colonIndex + 1).trim();

    if (key) {
      result[key] = value;
    }
  });

  return result;
}

/**
 * Parse a vulnerability patch TOON string into a ParsedVulnerability object
 * @param toonString - TOON formatted vulnerability string
 * @returns ParsedVulnerability object
 */
export function parseVulnerability(toonString: string): ParsedVulnerability {
  const data = parseTOON(toonString);

  // Extract line numbers if present
  let lineStart: number | undefined;
  let lineEnd: number | undefined;
  if (data.ln) {
    const lineParts = data.ln.split('-');
    lineStart = parseInt(lineParts[0], 10);
    lineEnd = lineParts.length > 1 ? parseInt(lineParts[1], 10) : lineStart;
  }

  // Generate deterministic ID based on content
  const idParts = [
    'vuln',
    data.file || 'unknown',
    lineStart || '0',
    data.category || 'unknown',
  ].join('-').replace(/[^a-zA-Z0-9-]/g, '_');
  const id = idParts;

  return {
    id,
    type: 'vulnerability',
    explanation: data.explanation || '',
    formattedCode: data.formatted_code || '',
    category: data.category || 'Unknown',
    severity: (data.sev || data.severity || 'MEDIUM').toUpperCase(),
    file: data.file,
    lineStart,
    lineEnd,

    // CVSS-like metrics
    attackVector: data.attack_vector,
    attackComplexity: data.attack_complexity,
    privilegesRequired: data.privileges_required,
    userInteraction: data.user_interaction,
    impactConfidentiality: data.impact_confidentiality,
    impactIntegrity: data.impact_integrity,
    impactAvailability: data.impact_availability,

    // Keep raw data for debugging
    raw: data,
  };
}

/**
 * Parse a secret patch TOON string into a ParsedVulnerability object
 * @param toonString - TOON formatted secret string
 * @returns ParsedVulnerability object (using same type for consistency)
 */
export function parseSecret(toonString: string): ParsedVulnerability {
  const data = parseTOON(toonString);

  // Extract line numbers if present
  let lineStart: number | undefined;
  let lineEnd: number | undefined;
  if (data.ln) {
    const lineParts = data.ln.split('-');
    lineStart = parseInt(lineParts[0], 10);
    lineEnd = lineParts.length > 1 ? parseInt(lineParts[1], 10) : lineStart;
  }

  // Generate deterministic ID based on content
  const idParts = [
    'secret',
    data.file || 'unknown',
    lineStart || '0',
    data.type || data.secret_type || 'unknown',
  ].join('-').replace(/[^a-zA-Z0-9-]/g, '_');
  const id = idParts;

  return {
    id,
    type: 'secret',
    explanation: data.explanation || '',
    formattedCode: data.formatted_code || '',
    file: data.file,
    lineStart,
    lineEnd,
    severity: (data.sev || data.severity || 'HIGH').toUpperCase(),

    // Secret-specific fields
    fixType: data.fix_type,
    recommendation: data.recommendation,
    secretType: data.type || data.secret_type,
    category: data.category || 'Hardcoded Secret',

    // Keep raw data for debugging
    raw: data,
  };
}

/**
 * Parse all vulnerabilities and secrets from scan data
 * @param vulnerabilities - Array of TOON formatted vulnerability strings
 * @param secrets - Array of TOON formatted secret strings
 * @returns Combined array of ParsedVulnerability objects
 */
export function parseAllFindings(
  vulnerabilities: string[],
  secrets: string[]
): ParsedVulnerability[] {
  const parsedVulns = vulnerabilities
    .filter((v) => v && v.trim())
    .map((v) => {
      try {
        return parseVulnerability(v);
      } catch (error) {
        console.warn('Failed to parse vulnerability:', v, error);
        return null;
      }
    })
    .filter((v): v is ParsedVulnerability => v !== null);

  const parsedSecrets = secrets
    .filter((s) => s && s.trim())
    .map((s) => {
      try {
        return parseSecret(s);
      } catch (error) {
        console.warn('Failed to parse secret:', s, error);
        return null;
      }
    })
    .filter((s): s is ParsedVulnerability => s !== null);

  return [...parsedVulns, ...parsedSecrets];
}

/**
 * Group vulnerabilities by file path
 * @param vulnerabilities - Array of ParsedVulnerability objects
 * @returns Map of file path to vulnerabilities
 */
export function groupVulnerabilitiesByFile(
  vulnerabilities: ParsedVulnerability[]
): Map<string, ParsedVulnerability[]> {
  const fileMap = new Map<string, ParsedVulnerability[]>();

  vulnerabilities.forEach((vuln) => {
    const filePath = vuln.file || 'unknown';
    if (!fileMap.has(filePath)) {
      fileMap.set(filePath, []);
    }
    fileMap.get(filePath)!.push(vuln);
  });

  // Sort vulnerabilities within each file by line number
  fileMap.forEach((vulns) => {
    vulns.sort((a, b) => {
      const lineA = a.lineStart || 0;
      const lineB = b.lineStart || 0;
      return lineA - lineB;
    });
  });

  return fileMap;
}

/**
 * Get vulnerability at a specific line number
 * @param vulnerabilities - Array of ParsedVulnerability objects
 * @param lineNumber - Line number to check
 * @returns Vulnerability at that line, or undefined
 */
export function getVulnerabilityAtLine(
  vulnerabilities: ParsedVulnerability[],
  lineNumber: number
): ParsedVulnerability | undefined {
  return vulnerabilities.find((vuln) => {
    if (!vuln.lineStart) return false;
    const start = vuln.lineStart;
    const end = vuln.lineEnd || start;
    return lineNumber >= start && lineNumber <= end;
  });
}
